_transcripts/010-rust-1.40.0.md

```diff
@@ -0,0 +1,748 @@
+`macro_rules!`, macros and procedural macros. And so, an `extern` block is, if
+you've ever used `extern "C" fn` before in Rust, as a way of defining the
+signature of a C function. `extern` blocks are a way of defining those kind of
+signatures, for many functions at once, as a convenient way, and for whatever
+reason, which I have— I don't understand right now, I need to actually ask—
+`macro_rules!` macros wouldn't expand within these blocks. Maybe it was just
+being a defensive kind of thing. But as of 1.40, `macro_rules!` macros now can
+be used inside of here. And you can also do things like put attributes on these
+`extern` `fn` definitions. So that's, again, just another place in Rust code
+that is no longer special, things should just work as you expect. So that's a
+nice change.
+
+__Jon__: I think the next one falls in a very similar category too, of,
+previously, you couldn't use `macro_rules!` to define a macro inside the output
+of a procedural macro, whereas now you can. Right? It's just another instance
+of, like, this wasn't possible before. And this is a place where macros had to
+be treated specialty. And now that is no longer the case.
+
+This last one, though, I think, is interesting, because it's a little bit odd
+when you first look at it. So when you use `macro_rules!` to define a macro that
+takes parameters, you can say what the type of those parameters should be, and
+the types here are a little bit different from normal, like, function parameter
+types, in that they are more types of expressions, rather than actual Rust
+types. So, for example, you can say that the first argument to a Rust macro
+should be a path to a type, or an expression, or an entire token tree, or a
+constant. So these are sort of, if you will, meta-level Rust constructs. You can
+say, I want this type of syntax, in this place, of a parameter for my macro, and
+one of those types is the `meta` type. The `meta` matcher, as it's called. And
+the `meta` matcher matches anything that looks like a attribute-style macro. So
+these would be things like `derive`s, for example, or the `serde::Serialize`
+attribute macros we talked about before, and the change that's happened here is
+that now that matcher accepts basically arbitrary sequences of tokens. So that
+is, you can— if you use that matcher, it will now accept basically arbitrary
+Rust code.
+
+And the reason for this change is a little interesting. Originally, the `meta`
+matcher was there because you wanted something that matched the syntax of these
+attribute-like macros, but over time as these types of macros have developed,
+the Rust developers have decided that they basically want to allow almost
+arbitrary syntax to be used in those attribute macros. Previously it would just
+like, path, brackets, and then attributes. Whereas now, they're leaning more
+towards, you should allow any kind of Rust syntax in there. But that also means
```

Comment:
is "path, brackets, and then attributes" some kind of syntax that should be in backticks?  I'm not sure what this is referring to.

Reply:
Ah, this refers to `#[path(attribute)]`. `path` here is a reference to the "token type" path, which you can use as, say, an AST type for a `macro_rules` argument, and it's basically either `foo` or `foo::bar::baz`. Not sure if backticks are better here?

Reply:
I think the backticks are at least a little helpful.  Even though it's not meant to be taken literally it's a bit more clear what's being discussed.